### DISCUSSION TITLE
Training Doesn't Utilize Available GPU/GPU Not Initialized

### DISCUSSION THREAD

**Discussion by jmdelahanty** [talmolab/sleap#904] [Help!] [answered]
Hello sleapyheads!

I was training a model on a cool computer with Fergil that has a nice Titan RTX on board but, while watching what the computer was up to during training, we saw that the GPU utilization was chilling around 0-5% much of the time! On the CPU side we saw the values going from around 15% to 100% in a cycle, probably in sync with batches that were running through the model.

Things were all set on a single animal model with default settings *except* for the batch size which we conservatively raised to 12 (from the default of 4). When looking at the terminal's outputs while the training was getting setup, we also noticed that while it said the GPU was available, it had not been initialized!

This is the kind of thing we were wondering about:

```
INFO:sleap.nn.training:
INFO:sleap.nn.training:Using GPU 0 for acceleration.
INFO:sleap.nn.training:Disabled GPU memory pre-allocation.
INFO:sleap.nn.training:System:
GPUs: 1/1 available
Device: /physical_device:GPU:0
Available: True
Initalized: False
Memory growth: True
```

~The `Initialized: False` statement was something we were wondering about. Is this pointing out that it's not going to use the GPU even if it's available?~ 

~Is this typical behavior when running a training? Is a card not initialized until the training itself actually starts?~

- Yes! The GPU is not initialized until training actually starts.

**Comment by talmo** [comment, marked as answer]
Hi @jmdelahanty,

GPU usage can fluctuate a lot and depending on the type of image and model, it may be that you're more bottlenecked by I/O or CPU than GPU operations.

Also, if you're on Windows, you definitely want to check that you're looking at the **Cuda** metric in the task manager:
![image](https://user-images.githubusercontent.com/3187454/184225770-c4717d53-13a0-478f-a273-d71003e2deab.png)

The `Initialized: False` is ok -- it won't get initialized until you actually start training. As long as it's available and you don't get any other errors in the console afterwards, you should be good.

Talmo

**Reply by jmdelahanty** [reply]
Will try running it while watching CUDA and report back! I bet that's what it was.

**Reply by jmdelahanty** [reply]
That's what it was! Thanks Talmo, didn't know that Windows had so many different metrics it could show for the GPU.

**Reply by roaldarbol** [reply]
In case anyone else comes by this, `CUDA` doesn't seem to be shown as an option in the Task Manager any longer - instead you need to look at `Compute 0`. :-)
